lib/streamlit/runtime/scriptrunner/script_requests.py

```diff
@@ -161,30 +176,36 @@ def request_rerun(self, new_data: RerunData) -> bool:
     def on_scriptrunner_yield(self) -> ScriptRequest | None:
         """Called by the ScriptRunner when it's at a yield point.
 
-        If we have no request or a RERUN request corresponding to one or more fragments,
-        return None.
+        If we have no request or a RERUN request corresponding to one or more fragments
+        (that is not a fragment-scoped rerun), return None.
 
-        If we have a (full script) RERUN request, return the request and set our internal
-        state to CONTINUE.
+        If we have a (full script or fragment-scoped) RERUN request, return the request
+        and set our internal state to CONTINUE.
 
         If we have a STOP request, return the request and remain stopped.
         """
         if self._state == ScriptRequestType.CONTINUE or (
-            # Reruns corresponding to fragments should *not* cancel the current script
-            # run as doing so will affect elements outside of the fragment.
+            # Reruns corresponding to fragment runs (that weren't caused by calls to
+            # `st.rerun(scope="fragment")`) should *not* cancel the current script run
+            # as doing so will affect elements outside of the fragment.
             self._state == ScriptRequestType.RERUN
             and self._rerun_data.fragment_id_queue
+            and not self._rerun_data.fragment_scoped_rerun
```

Comment:
Done! I didn't pull out the `state == ScriptRequestType.RERUN` since that'd make it get more than it has to be. I think the code ends up being roughly the same length, but it's nice having it be more clear that we're making the same check.